src/ray/raylet/agent_manager.cc

```diff
@@ -0,0 +1,93 @@
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/raylet/agent_manager.h"
+
+#include <thread>
+
+#include "ray/common/ray_config.h"
+#include "ray/util/logging.h"
+#include "ray/util/process.h"
+
+namespace ray {
+namespace raylet {
+
+void AgentManager::HandleRegisterAgent(const rpc::RegisterAgentRequest &request,
+                                       rpc::RegisterAgentReply *reply,
+                                       rpc::SendReplyCallback send_reply_callback) {
+  agent_ip_address_ = request.agent_ip_address();
+  agent_port_ = request.agent_port();
+  agent_pid_ = request.agent_pid();
+  RAY_LOG(INFO) << "HandleRegisterAgent, ip: " << agent_ip_address_
+                << ", port: " << agent_port_ << ", pid: " << agent_pid_;
+  reply->set_status(rpc::AGENT_RPC_STATUS_OK);
+  send_reply_callback(ray::Status::OK(), nullptr, nullptr);
+}
+
+void AgentManager::StartAgent() {
+  if (options_.agent_commands.empty()) {
+    RAY_LOG(INFO) << "Not to start agent, the agent command is empty.";
+    return;
+  }
+
+  if (RAY_LOG_ENABLED(DEBUG)) {
+    std::stringstream stream;
+    stream << "Starting agent process with command:";
+    for (const auto &arg : options_.agent_commands) {
+      stream << " " << arg;
+    }
+    RAY_LOG(DEBUG) << stream.str();
+  }
+
+  // Launch the process to create the agent.
+  std::error_code ec;
+  std::vector<const char *> argv;
+  for (const std::string &arg : options_.agent_commands) {
+    argv.push_back(arg.c_str());
+  }
+  argv.push_back(NULL);
+  Process child(argv.data(), nullptr, ec);
+  if (!child.IsValid() || ec) {
+    // The worker failed to start. This is a fatal error.
+    RAY_LOG(FATAL) << "Failed to start agent with return value " << ec << ": "
+                   << ec.message();
+  }
+
+  std::thread monitor_thread([this, child]() {
+    RAY_LOG(INFO) << "Monitor agent process with pid " << child.GetId()
+                  << ", register timeout "
+                  << RayConfig::instance().agent_register_timeout_ms() << "ms.";
+    auto timer = delay_executor_(
+        [this, child]() {
+          if (agent_pid_ != child.GetId()) {
+            RAY_LOG(WARNING) << "Agent process with pid " << child.GetId()
+                             << " has not registered, restart it.";
+            child.Kill();
+          }
+        },
+        RayConfig::instance().agent_register_timeout_ms());
+
+    int exit_code = child.Wait();
+    timer->cancel();
+
+    RAY_LOG(WARNING) << "Agent process with pid " << child.GetId()
+                     << " exit, return value " << exit_code;
+    RAY_UNUSED(delay_executor_([this] { StartAgent(); },
```

Comment:
We should kill this thread in this case?

Reply:
The `monitor_thread` has been detached. So, the thread will be exit after this lambda returns.

Reply:
Ah, that makes sense. I was confused.